File: apps/android/spectrocap-android/app/src/main/java/com/scingular/spectrocap/crypto/Format.kt
package com.scingular.spectrocap.crypto

import android.util.Base64
import com.google.gson.Gson
import com.google.gson.JsonObject
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Utilities for Phase 2A blob format and canonical metadata JSON.
 *
 * Phase 2A Blob Format:
 * [Magic (6 bytes: "SCAP2A")] + [Nonce (24 bytes)] + [Ciphertext (variable)]
 */
object BlobFormat {
    private const val MAGIC_STRING = "SCAP2A"
    private const val MAGIC_BYTES = 6
    private const val NONCE_SIZE = 24
    private val gson = Gson()

    /**
     * Creates Phase 2A blob from components.
     *
     * @param nonce 24-byte nonce
     * @param ciphertext Encrypted payload
     * @return Complete blob (magic + nonce + ciphertext)
     */
    fun createBlob(nonce: ByteArray, ciphertext: ByteArray): ByteArray {
        require(nonce.size == NONCE_SIZE) { "Nonce must be 24 bytes" }

        val totalSize = MAGIC_BYTES + NONCE_SIZE + ciphertext.size
        val blob = ByteArray(totalSize)

        // Write magic
        System.arraycopy(MAGIC_STRING.toByteArray(Charsets.UTF_8), 0, blob, 0, MAGIC_BYTES)

        // Write nonce
        System.arraycopy(nonce, 0, blob, MAGIC_BYTES, NONCE_SIZE)

        // Write ciphertext
        System.arraycopy(ciphertext, 0, blob, MAGIC_BYTES + NONCE_SIZE, ciphertext.size)

        return blob
    }

    /**
     * Parses Phase 2A blob into components.
     *
     * @param blob Complete Phase 2A blob
     * @return Pair of (nonce, ciphertext), or null if parsing fails
     */
    fun parseBlob(blob: ByteArray): Pair<ByteArray, ByteArray>? {
        return try {
            require(blob.size >= MAGIC_BYTES + NONCE_SIZE) { "Blob too short" }

            // Verify magic
            val magic = blob.sliceArray(0 until MAGIC_BYTES).toString(Charsets.UTF_8)
            require(magic == MAGIC_STRING) { "Invalid magic: $magic" }

            // Extract nonce
            val nonce = blob.sliceArray(MAGIC_BYTES until MAGIC_BYTES + NONCE_SIZE)

            // Extract ciphertext
            val ciphertext = blob.sliceArray(MAGIC_BYTES + NONCE_SIZE until blob.size)

            Pair(nonce, ciphertext)
        } catch (e: Exception) {
            null
        }
    }
}

/**
 * Utilities for canonical metadata JSON (Phase 2A).
 *
 * Canonical format ensures stable hash computation by maintaining alphabetical key order.
 */
object CanonicalMetadata {

    private val gson = Gson()

    /**
     * Creates canonical metadata JSON object with fields in alphabetical order.
     *
     * Order (MUST be alphabetical):
     * - alg
     * - createdAtClient
     * - messageId
     * - mime
     * - recipients (sorted array)
     * - senderDeviceId
     * - sizeBytesPlain
     * - storagePath
     * - type
     * - version
     *
     * @param messageId Message UUID
     * @param senderDeviceId Sender device UUID
     * @param recipients List of recipient device UUIDs (will be sorted)
     * @param storagePath Cloud Storage path (users/{uid}/messages/{messageId}.bin)
     * @param sizeBytesPlain Size of plaintext in bytes
     * @param createdAtClient ISO8601 timestamp of client creation
     * @param mimeType MIME type (typically "application/octet-stream" for Phase 2A)
     * @return Canonical JSON string (deterministic)
     */
    fun createCanonicalJson(
        messageId: String,
        senderDeviceId: String,
        recipients: List<String>,
        storagePath: String,
        sizeBytesPlain: Int,
        createdAtClient: String,
        mimeType: String = "application/octet-stream"
    ): String {
        // Create ordered map manually to ensure alphabetical key order
        val metadata = linkedMapOf<String, Any?>()

        metadata["alg"] = "xchacha20poly1305+sealedbox-x25519+ed25519"
        metadata["createdAtClient"] = createdAtClient
        metadata["messageId"] = messageId
        metadata["mime"] = mimeType
        metadata["recipients"] = recipients.sorted() // Sorted for stability
        metadata["senderDeviceId"] = senderDeviceId
        metadata["sizeBytesPlain"] = sizeBytesPlain
        metadata["storagePath"] = storagePath
        metadata["type"] = "text"
        metadata["version"] = "2A"

        return gson.toJson(metadata)
    }

    /**
     * Parses canonical metadata JSON (for verification on receiver side).
     *
     * @param jsonString Canonical metadata JSON
     * @return Map of metadata fields, or null if parsing fails
     */
    fun parseCanonical(jsonString: String): Map<String, Any?>? {
        return try {
            gson.fromJson(jsonString, Map::class.java) as Map<String, Any?>
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Computes metaHash = SHA256(canonicalJson).
     *
     * @param canonicalJson Canonical metadata JSON string
     * @return metaHash as ByteArray (32 bytes)
     */
    fun computeMetaHash(canonicalJson: String): ByteArray {
        return CryptoManager.sha256(canonicalJson.toByteArray(Charsets.UTF_8))
    }

    /**
     * Verifies metaHash by recomputing and comparing.
     *
     * @param canonicalJson Canonical metadata JSON string
     * @param metaHashBase64 Expected metaHash (base64)
     * @return true if hashes match
     */
    fun verifyMetaHash(canonicalJson: String, metaHashBase64: String): Boolean {
        val computed = computeMetaHash(canonicalJson)
        val expected = Base64.decode(metaHashBase64, Base64.NO_WRAP)
        return computed.contentEquals(expected)
    }
}

File: apps/windows/src/crypto/format.rs
/// Phase 2A blob format and canonical JSON utilities

use serde_json::{json, Value};
use super::primitives::CryptoPrimitives;

pub struct BlobFormat;

impl BlobFormat {
    const MAGIC: &'static [u8] = b"SCAP2A";
    const MAGIC_LEN: usize = 6;
    const NONCE_LEN: usize = 24;

    /// Creates Phase 2A blob from components
    /// 
    /// Format: [Magic (6 bytes)] + [Nonce (24 bytes)] + [Ciphertext (variable)]
    /// 
    /// # Arguments
    /// * `nonce` - 24-byte nonce
    /// * `ciphertext` - Encrypted payload with AEAD tag
    /// 
    /// # Returns
    /// Complete blob
    pub fn create_blob(nonce: &[u8], ciphertext: &[u8]) -> Result<Vec<u8>, String> {
        if nonce.len() != Self::NONCE_LEN {
            return Err(format!("Nonce must be {} bytes", Self::NONCE_LEN));
        }

        let mut blob = Vec::with_capacity(Self::MAGIC_LEN + Self::NONCE_LEN + ciphertext.len());
        blob.extend_from_slice(Self::MAGIC);
        blob.extend_from_slice(nonce);
        blob.extend_from_slice(ciphertext);

        Ok(blob)
    }

    /// Parses Phase 2A blob into components
    /// 
    /// # Returns
    /// Tuple of (nonce, ciphertext), or error if parsing fails
    pub fn parse_blob(blob: &[u8]) -> Result<(Vec<u8>, Vec<u8>), String> {
        if blob.len() < Self::MAGIC_LEN + Self::NONCE_LEN {
            return Err("Blob too short".to_string());
        }

        // Verify magic
        if &blob[0..Self::MAGIC_LEN] != Self::MAGIC {
            return Err("Invalid blob magic".to_string());
        }

        let nonce = blob[Self::MAGIC_LEN..Self::MAGIC_LEN + Self::NONCE_LEN].to_vec();
        let ciphertext = blob[Self::MAGIC_LEN + Self::NONCE_LEN..].to_vec();

        Ok((nonce, ciphertext))
    }
}

/// Canonical metadata for Phase 2A (stable JSON key order)
pub struct CanonicalMetadata;

impl CanonicalMetadata {
    /// Creates canonical metadata JSON (alphabetical key order)
    /// 
    /// Order (ALPHABETICAL):
    /// - alg
    /// - createdAtClient
    /// - messageId
    /// - mime
    /// - recipients (sorted array)
    /// - senderDeviceId
    /// - sizeBytesPlain
    /// - storagePath
    /// - type
    /// - version
    pub fn create_canonical_json(
        message_id: &str,
        sender_device_id: &str,
        recipients: &[String],
        storage_path: &str,
        size_bytes_plain: usize,
        created_at_client: &str,
    ) -> String {
        let mut sorted_recipients = recipients.to_vec();
        sorted_recipients.sort();

        // Build in alphabetical order
        let mut map = serde_json::Map::new();
        map.insert("alg".to_string(), json!("xchacha20poly1305+sealedbox-x25519+ed25519"));
        map.insert("createdAtClient".to_string(), json!(created_at_client));
        map.insert("messageId".to_string(), json!(message_id));
        map.insert("mime".to_string(), json!("application/octet-stream"));
        map.insert("recipients".to_string(), json!(sorted_recipients));
        map.insert("senderDeviceId".to_string(), json!(sender_device_id));
        map.insert("sizeBytesPlain".to_string(), json!(size_bytes_plain));
        map.insert("storagePath".to_string(), json!(storage_path));
        map.insert("type".to_string(), json!("text"));
        map.insert("version".to_string(), json!("2A"));

        serde_json::to_string(&Value::Object(map)).expect("Failed to serialize")
    }

    /// Computes metaHash = SHA256(canonicalJson)
    pub fn compute_meta_hash(canonical_json: &str) -> Vec<u8> {
        CryptoPrimitives::sha256(canonical_json.as_bytes())
    }

    /// Verifies metaHash by recomputing and comparing
    pub fn verify_meta_hash(canonical_json: &str, meta_hash_expected: &[u8]) -> bool {
        let computed = Self::compute_meta_hash(canonical_json);
        computed == meta_hash_expected
    }

    /// Reconstructs canonical JSON from Firestore document
    /// (for verification on receiver side)
    pub fn from_firestore_doc(doc: &Value) -> Result<String, String> {
        let message_id = doc.get("messageId")
            .and_then(|v| v.as_str())
            .ok_or("Missing messageId")?;
        
        let sender_device_id = doc.get("senderDeviceId")
            .and_then(|v| v.as_str())
            .ok_or("Missing senderDeviceId")?;
        
        let recipients = doc.get("recipients")
            .and_then(|v| v.as_array())
            .ok_or("Missing recipients")?
            .iter()
            .filter_map(|v| v.as_str().map(String::from))
            .collect::<Vec<_>>();
        
        let storage_path = doc.get("storagePath")
            .and_then(|v| v.as_str())
            .ok_or("Missing storagePath")?;
        
        let size_bytes_plain = doc.get("sizeBytesPlain")
            .and_then(|v| v.as_u64())
            .ok_or("Missing sizeBytesPlain")? as usize;
        
        let created_at_client = doc.get("createdAtClient")
            .and_then(|v| v.as_str())
            .ok_or("Missing createdAtClient")?;

        Ok(Self::create_canonical_json(
            message_id,
            sender_device_id,
            &recipients,
            storage_path,
            size_bytes_plain,
            created_at_client,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_canonical_json_alphabetical() {
        let json = CanonicalMetadata::create_canonical_json(
            "msg-123",
            "dev-456",
            &["dev-456".to_string(), "dev-789".to_string()],
            "users/uid/messages/msg-123.bin",
            100,
            "2026-01-28T16:45:00Z",
        );

        // Verify it contains expected fields in order
        assert!(json.contains("\"alg\""));
        assert!(json.contains("\"version\""));
        
        // Alphabetical order check via JSON parsing
        let parsed: serde_json::Map<String, Value> = serde_json::from_str(&json)
            .expect("Valid JSON");
        let keys: Vec<_> = parsed.keys().cloned().collect();
        let mut sorted_keys = keys.clone();
        sorted_keys.sort();
        assert_eq!(keys, sorted_keys, "Keys not in alphabetical order");
    }
}
